server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -4324,12 +4458,35 @@ public void resetEngine() {
         assert Thread.holdsLock(mutex) == false : "resetting engine under mutex";
         assert waitForEngineOrClosedShardListeners.isDone();
         try {
-            synchronized (engineMutex) {
+            engineLock.readLock().lock();
+            var release = true;
+            try {
                 verifyNotClosed();
-                getEngine().prepareForEngineReset();
-                var newEngine = createEngine(newEngineConfig(replicationTracker));
-                IOUtils.close(currentEngineReference.getAndSet(newEngine));
-                onNewEngine(newEngine);
+                // Preparing the engine for reset flushes and blocks for refresh: it should not be executed under the write lock because
+                // another thread might already be refreshing, and the refresh listeners in that thread will need to access to the engine
+                // using the read lock. If we were using the write lock here, it would deadlock.
+                currentEngine.prepareForEngineReset();
+                engineLock.readLock().unlock();
+                release = false;
+
+                // Promote to write lock in order to swap engines
+                engineLock.writeLock().lock();
+                try {
+
+                    // How do we ensure that no indexing operations have been processed since prepareForEngineReset() here? We're not
+                    // blocking all operations when resetting the engine nor we are blocking flushes or force-merges.
+
+                    assert state != IndexShardState.CLOSED : "state has changed without holding engine write lock!";
+                    var newEngine = createEngine(newEngineConfig(replicationTracker));
+                    IOUtils.close(getAndSetCurrentEngine(newEngine));
```

Comment:
I wonder if we should close the previous engine after releasing the lock?

Reply:
I think we could, but closing the engine while holding the write lock might help us for dynamic hollowing I think.

See my https://github.com/elastic/elasticsearch/pull/124635#discussion_r2007094141

Reply:
This is a good point, and I think that today nothing enforces the right ordering for these, right? We can take care of that in a follow-up.